### PR TITLE
Copy SVD implementation from Unity.Animation

### PR DESCRIPTION
**Pull request by unpacklo**
Copied over the SVD implementation so that way we can expose the pseudoinverse and `rotation` methods.

Most code is copy/paste except for the tests which needed to have different asserts so Burst can run the tests in their repository.

### CODE REVIEW

File: src/Tests/Tests/Shared/TestQuaternion.cs [hunk 1]
```diff
@@ -380,5 +380,85 @@ public static void quaternion_angle()
...
+        public static void quaternion_rotation_from_3x3_with_negative_nonuniform_scale()
+        {
+            const float tolerance = 1e-5f;
+            var random = new Random(912u);
+            var expectedQuaternion = random.NextQuaternionRotation();
+            var m = new float3x3(expectedQuaternion);
+            m = mul(m, float3x3.Scale(-1.5f, -0.2f, -51.281f));
+            var actualQuaternion = rotation(m);
+            TestUtils.AreEqual(0.0f, angle(actualQuaternion, expectedQuaternion) % PI * 2.0f, tolerance);
+        }
```
Comment (mecanimdev) [inline src/Tests/Tests/Shared/TestQuaternion.cs:462]: we should also add two more tests:

one with very small scale (1e-12f, 1e-12f, 1e-12f)
and one with zero scale (1,1,0)

**Review by mecanimdev — COMMENTED** [review]
lgtm

**Review by oldio — APPROVED** [review]
LGTM

A test similar to this would be good as Sonny mentioned :
https://github.cds.internal.unity3d.com/unity/dots/blob/5e6042e3d299482f6f26defac24350ac1f43db61/Packages/com.unity.animation/Unity.Animation.Tests/AffineTransformTests.cs#L58

File: src/Unity.Mathematics/quaternion.cs [hunk 2]
```diff
@@ -710,6 +710,50 @@ public static float angle(quaternion q1, quaternion q2)
             return diff + diff;
         }
 
+        /// <summary>
+        /// Extracts the rotation from a matrix that potentially has scale.
```
Comment (jo-unity) [inline src/Unity.Mathematics/quaternion.cs:714]: "that potentially has scale."

What do you mean by "potentially"? If it has a scale then I assume this extracts the rotation. However, if the matrix _doesn't_ have a scale, what happens?

I'd rephrase this to something like:

Extracts the rotation from a matrix that has a a scale. If the matrix doesn't have a scale then [thing happens].

Presumably returns null or something?

Comment (mecanimdev) [reply src/Unity.Mathematics/quaternion.cs:714]: this method is the generic form that can support any type of rotation matrix. With no scale, uniform scale, non uniform scale, non singular rotation matrix. It will always returns a valid rotation matrix.

All other methods to extract rotation can fail silently to some degree depending on the input matrix configuration.

Comment (jo-unity) [reply src/Unity.Mathematics/quaternion.cs:714]: Ah okay, in that case, I don't understand why we need to mention that it potentially has scale, if it generally supports all types

Comment (mecanimdev) [reply src/Unity.Mathematics/quaternion.cs:714]: well ... the thing is how do we communicate to our users that if you have non uniform scale in your rotation matrix you need to use this method, all others one will fail and yield bad result

Comment (jo-unity) [reply src/Unity.Mathematics/quaternion.cs:714]: Ah ok, well, I feel like we could probably say just that - it seems rather important! my main feedback is that it's currently worded vaguely and we could be more explicit about the fact that you should use this in specific situations e.g 

"Extracts the rotation from a matrix. This method supports any type of rotation matrix: if the matrix has a non uniform scale you should use this method."

File: src/Tests/Tests/Shared/TestQuaternion.cs (hunk 1 above)
Comment (unpacklo) [reply src/Tests/Tests/Shared/TestQuaternion.cs:462]: Sigh, I thought the tiny scale test I added was fine but there's actually a difference between the results when I run the test directly from the solution vs running the test with Unity. Unity generates NaNs during one of the jacobi iterations, so I'm digging into that now.

Comment (unpacklo) [reply src/Tests/Tests/Shared/TestQuaternion.cs:462]: I see the difference now. The existing picoscale test (https://github.cds.internal.unity3d.com/unity/dots/blob/5e6042e3d299482f6f26defac24350ac1f43db61/Packages/com.unity.animation/Unity.Animation.Tests/AffineTransformTests.cs#L58) actually uses `mathex.inverse` from Unity.Animation which rescales the input matrix. But the rotation method uses `svdRotation` which does no such thing and eventually the code runs into a normalization which causes the tiny values to underflow.

image

Do we need to handle these tiny scales in `svdRotation` to make this test pass?

File: src/Unity.Mathematics/quaternion.cs (hunk 2 above)
Comment (unpacklo) [reply src/Unity.Mathematics/quaternion.cs]: @jo-unity I updated the docs to use the text you provided above. I split them apart into the summary and remarks sections, hopefully that's ok?

File: src/Tests/Tests/Shared/TestQuaternion.cs (hunk 1 above)
Comment (mecanimdev) [reply src/Tests/Tests/Shared/TestQuaternion.cs:462]: you're right, it the affine transform from UnityEngine that does have a test with pico scale, not the sdvRotation.

So we are on par with what we have in Unity so we can drop the pico scale test on svd.
